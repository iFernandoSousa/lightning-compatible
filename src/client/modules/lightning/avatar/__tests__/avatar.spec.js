/*
 * Copyright (c) 2019, salesforce.com, inc.
 * All rights reserved.
 * SPDX-License-Identifier: MIT
 * For full license text, see the LICENSE file in the repo root or https://opensource.org/licenses/MIT
 */

import { createElement } from 'lwc';
import Element from 'c/avatar';
import { shadowQuerySelector } from 'lightning/testUtils';

const createAvatar = attributes => {
    const element = createElement('c-avatar', { is: Element });

    Object.assign(element, attributes);
    document.body.appendChild(element);
    return element;
};

describe('c-avatar', () => {
    it('default', () => {
        const element = createAvatar({
            alternativeText: 'required alternative text',
            fallbackIconName: 'standard:account'
        });

        return Promise.resolve().then(() => {
            expect(element).toMatchSnapshot();
        });
    });

    describe('testFallbackIconName', () => {
        it('lightning:icon should render', () => {
            const element = createAvatar({
                alternativeText: 'required alternative text',
                fallbackIconName: 'standard:account',
                initials: '',
                src: ''
            });

            return Promise.resolve().then(() => {
                const icon = shadowQuerySelector(element, 'c-icon');
                expect(icon).toBeTruthy();
            });
        });
    });

    describe('initials', () => {
        it('should not render when invalid', () => {
            const element = createAvatar({
                alternativeText: 'required alternative text',
                fallbackIconName: 'standard:account',
                initials: '',
                src: ''
            });

            return Promise.resolve().then(() => {
                const abbr = shadowQuerySelector(element, 'abbr');
                expect(abbr).toBeFalsy();
            });
        });

        it('should render when valid', () => {
            const element = createAvatar({
                alternativeText: 'required alternative text',
                fallbackIconName: 'standard:account',
                initials: 'EK',
                src: ''
            });

            return Promise.resolve().then(() => {
                const abbr = shadowQuerySelector(element, 'abbr');
                expect(abbr).toBeTruthy();
            });
        });

        it('should render class slds-icon-custom-custom79 when fallbackIconName is custom:custom79', () => {
            const element = createAvatar({
                alternativeText: 'required alternative text',
                fallbackIconName: 'custom:custom79',
                initials: 'EK',
                src: ''
            });

            return Promise.resolve().then(() => {
                const abbr = shadowQuerySelector(element, 'abbr');
                expect(abbr.className).toContain('slds-icon-custom-custom79');
            });
        });
    });

    describe('testSrcAttribute', () => {
        it('image src should be correct when changed', () => {
            const element = createAvatar({
                alternativeText: 'account owner',
                fallbackIconName: 'standard:account',
                initials: 'EK',
                src: '/assets/images/avatar2.jpg'
            });

            return Promise.resolve()
                .then(() => {
                    expect(element).toMatchSnapshot();
                    element.src = '/assets/images/avatar1.jpg';
                    return Promise.resolve();
                })
                .then(() => {
                    expect(element).toMatchSnapshot();
                });
        });

        it('An img element should not render when v.src is undefined', () => {
            const element = createAvatar({
                fallbackIconName: 'standard:account',
                alternativeText: 'account owner',
                initials: 'EK'
            });

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });
    });

    describe('testAltTitleAttributes', () => {
        it('Avatar component should set title of image correctly', () => {
            const element = createAvatar({
                alternativeText: 'account owner',
                fallbackIconName: 'standard:account',
                src: '/assets/images/avatar2.jpg'
            });

            return Promise.resolve()
                .then(() => {
                    expect(element).toMatchSnapshot();
                    element.alternativeText = 'Salesforce Contact';

                    return Promise.resolve();
                })
                .then(() => {
                    expect(element).toMatchSnapshot();
                });
        });
    });

    describe('Variant', () => {
        const createAvatarWithVariant = variant => {
            const element = createAvatar({
                alternativeText: 'account owner',
                fallbackIconName: 'standard:account',
                src: '/assets/images/avatar2.jpg',
                variant
            });

            return element;
        };

        it('should add default classes on invalid variant', () => {
            const element = createAvatarWithVariant('invalid');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('should add correct classes on circle variant', () => {
            const element = createAvatarWithVariant('circle');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('should add correct classes on square variant', () => {
            const element = createAvatarWithVariant('square');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });
    });

    describe('Size', () => {
        const createAvatarWithSize = size => {
            const element = createAvatar({
                alternativeText: 'account owner',
                fallbackIconName: 'standard:account',
                src: '/assets/images/avatar2.jpg',
                size
            });

            return element;
        };

        it('should add default classes on invalid size', () => {
            const element = createAvatarWithSize('invalid');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('should add correct classes on x-small size', () => {
            const element = createAvatarWithSize('x-small');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('should add correct classes on small size', () => {
            const element = createAvatarWithSize('small');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('should add correct classes on medium size', () => {
            const element = createAvatarWithSize('medium');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });

        it('should add correct classes on large size', () => {
            const element = createAvatarWithSize('large');

            return Promise.resolve().then(() => {
                expect(element).toMatchSnapshot();
            });
        });
    });

    it('testTypePriority', () => {
        const element = createAvatar({
            alternativeText: 'required alternative text',
            fallbackIconName: 'standard:account',
            src: '/assets/images/avatar2.jpg'
        });

        return Promise.resolve()
            .then(() => {
                expect(element).toMatchSnapshot();

                element.src = '';
                element.initials = 'EK';

                return Promise.resolve();
            })
            .then(() => {
                expect(element).toMatchSnapshot();

                element.src = '';
                element.initials = '';

                return Promise.resolve();
            })
            .then(() => {
                expect(element).toMatchSnapshot();

                element.src = '/assets/images/avatar1.jpg';
                element.initials = 'Xx';

                return Promise.resolve();
            })
            .then(() => {
                expect(element).toMatchSnapshot();
            });
    });

    it('testSrcAttribute using valid and invalid src', () => {
        const element = createAvatar({
            alternativeText: 'account owner',
            fallbackIconName: 'standard:account',
            initials: 'EK',
            src: '/assets/images/avatar2.jpg'
        });

        return Promise.resolve()
            .then(() => {
                const img = shadowQuerySelector(element, 'img');
                expect(img).not.toBeNull();
                expect(img.src).toBe(
                    'http://localhost/assets/images/avatar2.jpg'
                );

                element.src = '';
            })
            .then(() => {
                expect(shadowQuerySelector(element, 'img')).toBeNull();
                element.src = 'http://localhost/assets/images/avatar1.jpg';
            })
            .then(() => {
                const img = shadowQuerySelector(element, 'img');
                expect(img).not.toBeNull();
                expect(img.src).toBe(
                    'http://localhost/assets/images/avatar1.jpg'
                );
            });
    });
});
