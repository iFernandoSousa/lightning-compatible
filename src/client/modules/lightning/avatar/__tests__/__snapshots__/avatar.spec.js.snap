// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`c-avatar Size should add correct classes on large size 1`] = `
<c-avatar
  class="slds-avatar slds-avatar_large"
>
  #shadow-root(open)
    <img
      alt="account owner"
      src="/assets/images/avatar2.jpg"
      title="account owner"
    />
</c-avatar>
`;

exports[`c-avatar Size should add correct classes on medium size 1`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <img
      alt="account owner"
      src="/assets/images/avatar2.jpg"
      title="account owner"
    />
</c-avatar>
`;

exports[`c-avatar Size should add correct classes on small size 1`] = `
<c-avatar
  class="slds-avatar slds-avatar_small"
>
  #shadow-root(open)
    <img
      alt="account owner"
      src="/assets/images/avatar2.jpg"
      title="account owner"
    />
</c-avatar>
`;

exports[`c-avatar Size should add correct classes on x-small size 1`] = `
<c-avatar
  class="slds-avatar slds-avatar_x-small"
>
  #shadow-root(open)
    <img
      alt="account owner"
      src="/assets/images/avatar2.jpg"
      title="account owner"
    />
</c-avatar>
`;

exports[`c-avatar Size should add default classes on invalid size 1`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <img
      alt="account owner"
      src="/assets/images/avatar2.jpg"
      title="account owner"
    />
</c-avatar>
`;

exports[`c-avatar Variant should add correct classes on circle variant 1`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium slds-avatar_circle"
>
  #shadow-root(open)
    <img
      alt="account owner"
      src="/assets/images/avatar2.jpg"
      title="account owner"
    />
</c-avatar>
`;

exports[`c-avatar Variant should add correct classes on square variant 1`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <img
      alt="account owner"
      src="/assets/images/avatar2.jpg"
      title="account owner"
    />
</c-avatar>
`;

exports[`c-avatar Variant should add default classes on invalid variant 1`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <img
      alt="account owner"
      src="/assets/images/avatar2.jpg"
      title="account owner"
    />
</c-avatar>
`;

exports[`c-avatar default 1`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <c-icon
      class="slds-icon-standard-account slds-icon_container"
      title="required alternative text"
    >
      #shadow-root(open)
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-icon"
              data-key="account"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/standard-sprite/svg/symbols.svg#account"
              />
            </svg>
        </c-primitive-icon>
        <span
          class="slds-assistive-text"
        >
          required alternative text
        </span>
    </c-icon>
</c-avatar>
`;

exports[`c-avatar testAltTitleAttributes Avatar component should set title of image correctly 1`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <img
      alt="account owner"
      src="/assets/images/avatar2.jpg"
      title="account owner"
    />
</c-avatar>
`;

exports[`c-avatar testAltTitleAttributes Avatar component should set title of image correctly 2`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <img
      alt="Salesforce Contact"
      src="/assets/images/avatar2.jpg"
      title="Salesforce Contact"
    />
</c-avatar>
`;

exports[`c-avatar testSrcAttribute An img element should not render when v.src is undefined 1`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <abbr
      class="slds-avatar__initials slds-icon-standard-account"
      title="account owner"
    >
      EK
    </abbr>
</c-avatar>
`;

exports[`c-avatar testSrcAttribute image src should be correct when changed 1`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <img
      alt="account owner"
      src="/assets/images/avatar2.jpg"
      title="account owner"
    />
</c-avatar>
`;

exports[`c-avatar testSrcAttribute image src should be correct when changed 2`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <img
      alt="account owner"
      src="/assets/images/avatar1.jpg"
      title="account owner"
    />
</c-avatar>
`;

exports[`c-avatar testTypePriority 1`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <img
      alt="required alternative text"
      src="/assets/images/avatar2.jpg"
      title="required alternative text"
    />
</c-avatar>
`;

exports[`c-avatar testTypePriority 2`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <abbr
      class="slds-avatar__initials slds-icon-standard-account"
      title="required alternative text"
    >
      EK
    </abbr>
</c-avatar>
`;

exports[`c-avatar testTypePriority 3`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <c-icon
      class="slds-icon-standard-account slds-icon_container"
      title="required alternative text"
    >
      #shadow-root(open)
        <c-primitive-icon>
          #shadow-root(open)
            <svg
              aria-hidden="true"
              class="slds-icon"
              data-key="account"
              focusable="false"
            >
              <use
                xlink:href="/assets/icons/standard-sprite/svg/symbols.svg#account"
              />
            </svg>
        </c-primitive-icon>
        <span
          class="slds-assistive-text"
        >
          required alternative text
        </span>
    </c-icon>
</c-avatar>
`;

exports[`c-avatar testTypePriority 4`] = `
<c-avatar
  class="slds-avatar slds-avatar_medium"
>
  #shadow-root(open)
    <img
      alt="required alternative text"
      src="/assets/images/avatar1.jpg"
      title="required alternative text"
    />
</c-avatar>
`;
